trl/trainer/sft_trainer.py

```diff
@@ -0,0 +1,310 @@
+            args=args,
+            data_collator=data_collator,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            tokenizer=tokenizer,
+            model_init=model_init,
+            compute_metrics=compute_metrics,
+            callbacks=callbacks,
+            optimizers=optimizers,
+            preprocess_logits_for_metrics=preprocess_logits_for_metrics,
+        )
+
+    def _prepare_dataset(
+        self,
+        dataset,
+        tokenizer,
+        packing,
+        dataset_text_field,
+        max_seq_length,
+        formatting_func,
+        infinite,
+        num_of_sequences,
+        chars_per_token,
+    ):
+        # check if torch dataset / dataloader and do nothing
+        if dataset is not None and (
+            isinstance(
+                dataset,
+                (torch.utils.data.IterableDataset, torch.utils.data.Dataset),
+            )
+        ):
+            is_already_dataset = True
+        else:
+            is_already_dataset = False
+
+        if not packing:
+            dataset = self._prepare_non_packed_dataloader(
+                tokenizer, dataset, dataset_text_field, max_seq_length, formatting_func
+            )
+
+            is_already_dataset = True
+
+        if not is_already_dataset and (dataset_text_field is not None or formatting_func is not None):
+            if tokenizer is None:
+                raise ValueError(
+                    "You need to pass a tokenizer when using the SFT Trainer when passing a `dataset_text_field`."
+                )
+
+            dataset = ConstantLengthDataset(
+                tokenizer,
+                dataset[dataset_text_field] if dataset_text_field is not None else dataset,
+                formatting_func=formatting_func,
+                seq_length=max_seq_length,
+                infinite=infinite,
+                num_of_sequences=num_of_sequences,
+                chars_per_token=chars_per_token,
+                eos_token_id=tokenizer.eos_token_id,
+            )
+
+        elif not is_already_dataset and (dataset_text_field is None and formatting_func is None):
+            raise ValueError(
+                "You need to pass a `dataset_text_field` or `formatting_func` argument to the SFTTrainer if you want to use the `ConstantLengthDataset`."
+            )
+
+        return dataset
+
+    def _prepare_non_packed_dataloader(
+        self, tokenizer, dataset, dataset_text_field, max_seq_len, formatting_func=None
+    ):
+        use_formatting_func = formatting_func is not None and dataset_text_field is None
+
+        # Inspired from: https://huggingface.co/learn/nlp-course/chapter7/6?fw=pt
+        def tokenize(element):
+            outputs = tokenizer(
+                element[dataset_text_field] if not use_formatting_func else formatting_func(element),
+                truncation=True,
+                max_length=max_seq_len,
+                return_overflowing_tokens=True,
```

Comment:
I think we should probably set this to False. If you are not using packing then having batches with have samples is a bit odd. What do you think?

Reply:
Yes makes sense!